pkg/bitlpm/trie.go

```diff
@@ -0,0 +1,536 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package bitlpm
+
+// Trie is a non-thread-safe binary trie that indexes arbitrarily long
+// bit-based keys with associated prefix lengths indexed from [most significant bit]
+// ("MSB") to [least significant bit] ("LSB") using the
+// [longest prefix match algorithm].
+//
+// A prefix-length (hereafter "prefix"), in a prefix-key pair or in a RangeLookup, represents the
+// minimum number of bits (from MSB to LSB) that another comparable key
+// must match.
+//
+// Each method's comments describes the mechanism of how the method
+// works.
+//
+// [most significant bit]: https://en.wikipedia.org/wiki/Bit_numbering#Most_significant_bit
+// [least significant bit]: https://en.wikipedia.org/wiki/Bit_numbering#Least_significant_bit
+// [longest prefix match algorithm]: https://en.wikipedia.org/wiki/Longest_prefix_match
+type Trie[K, T any] interface {
+	// Lookup returns the longest prefix match to a specific
+	// key.
+	Lookup(key K) T
```

Comment:
Should this be dual-valued? It might be nice to know if no matches were found. Alternatively, `TryLookup(key K) (T, bool)`

Reply:
> Should this be dual-valued? It might be nice to know if no matches were found. Alternatively, TryLookup(key K) (T, bool)

If this becomes a problem we can always come back and fix it later (pretty easily). However, 9/10 times this is probably going to be a pointer and the default `nil` value will be returned when it fails.